clients/pkg/promtail/targets/gcplog/pull_target.go

```diff
@@ -83,42 +101,57 @@ func (t *pullTarget) run() error {
 	t.wg.Add(1)
 	defer t.wg.Done()
 
-	send := t.handler.Chan()
-
-	sub := t.ps.SubscriptionInProject(t.config.Subscription, t.config.ProjectID)
-	go func() {
-		// NOTE(kavi): `cancel` the context as exiting from this goroutine should stop main `run` loop
-		// It makesense as no more messages will be received.
-		defer t.cancel()
-
-		err := sub.Receive(t.ctx, func(ctx context.Context, m *pubsub.Message) {
-			t.msgs <- m
-		})
-		if err != nil {
-			level.Error(t.logger).Log("msg", "failed to receive pubsub messages", "error", err)
-			t.metrics.gcplogErrors.WithLabelValues(t.config.ProjectID).Inc()
-			t.metrics.gcplogTargetLastSuccessScrape.WithLabelValues(t.config.ProjectID, t.config.Subscription).SetToCurrentTime()
-		}
-	}()
+	subscriptionErr := make(chan error)
+	go t.consumeSubscription(subscriptionErr)
 
 	for {
 		select {
 		case <-t.ctx.Done():
 			return t.ctx.Err()
+		case e := <-subscriptionErr:
+			return e
 		case m := <-t.msgs:
 			entry, err := parseGCPLogsEntry(m.Data, t.config.Labels, nil, t.config.UseIncomingTimestamp, t.relabelConfig)
 			if err != nil {
 				level.Error(t.logger).Log("event", "error formating log entry", "cause", err)
 				m.Ack()
 				break
 			}
-			send <- entry
+			t.handler.Chan() <- entry
 			m.Ack() // Ack only after log is sent.
 			t.metrics.gcplogEntries.WithLabelValues(t.config.ProjectID).Inc()
 		}
 	}
 }
 
+func (t *pullTarget) consumeSubscription(subscriptionErr chan error) {
+	// NOTE(kavi): `cancel` the context as exiting from this goroutine should stop main `run` loop
+	// It makesense as no more messages will be received.
+	defer t.cancel()
+
+	var lastError error
+	for t.backoff.Ongoing() {
+		lastError = t.sub.Receive(t.ctx, func(ctx context.Context, m *pubsub.Message) {
+			t.msgs <- m
+
+			// When the subscription works properly, it doesn't return
+			// Reset relevant state here
+			lastError = nil
+			t.backoff.Reset()
+		})
+		if lastError != nil {
+			level.Error(t.logger).Log("msg", "failed to receive pubsub messages", "error", lastError)
+			t.metrics.gcplogErrors.WithLabelValues(t.config.ProjectID).Inc()
+			t.metrics.gcplogTargetLastSuccessScrape.WithLabelValues(t.config.ProjectID, t.config.Subscription).SetToCurrentTime()
+			t.backoff.Wait()
+		}
+	}
+
+	if t.ctx.Err() == nil && t.backoff.Err() != nil {
```

Comment:
hmm why do you want to skip this check when `t.ctx.Err()`?

Reply:
`backoff.Err()` returns the `ctx.Err()` if the context has been canceled. We handle the context-canceled case already so this ensures we're only handling errors from the pubsub